src/syntree.rs

```diff
@@ -189,6 +189,115 @@ pub struct Type {
     kind: TypeKind,
 }
 
-pub fn construct(tokens: TokenStream) -> Result<Prog, Vec<Error>> {
-    Err(Vec::new())
+type Tokens = [(Token, usize)];
+
+#[derive(Debug, Copy, Clone)]
+struct Context<'a> {
+    pub tokens: &'a Tokens,
+    pub curr: usize,
+    pub file: &'a Path,
+}
+
+impl<'a> Context<'a> {
+
+    fn new(tokens: &'a [(Token, usize)], file: &'a Path) -> Self {
+        Self { tokens, curr: 0, file, }
+    }
+
+    fn span(&self) -> Span {
+        Span { line: self.peek().1 }
+    }
+
+    fn line(&self) -> usize {
+        self.span().line
+    }
+
+    fn skip(&self, n: usize) -> Self {
+        let mut new = self.clone();
+        new.curr += n;
+        new
+    }
+
+    fn peek(&self) -> &(Token, usize) {
+        &self.tokens.get(self.curr).unwrap_or(&(Token::EOF, 0))
+    }
+
+    fn token(&self) -> &Token {
+        &self.peek().0
+    }
+
+    fn eat(&self) -> (&Token, Span, Context) {
+        (self.token(), self.span(), self.skip(1))
+    }
+}
+
+macro_rules! syntax_error {
+    ($ctx:expr, $( $msg:expr ),* ) => {
+        {
+            let msg = format!($( $msg ),*).into();
+            Error::SyntaxError {
+                file: $ctx.file.to_path_buf(),
+                line: $ctx.line(),
+                token: $ctx.token().clone(),
+                message: Some(msg),
+            }
+        }
+    };
+}
+
+macro_rules! raise_syntax_error {
+    ($ctx:expr, $( $msg:expr ),* ) => {
+        return Err(($ctx.skip(1), vec![syntax_error!($ctx, $( $msg ),*)]))
+    };
+}
+
+fn expression<'t>(ctx: Context<'t>) -> Result<(Context<'t>, Expression), (Context<'t>, Vec<Error>)> {
```

Comment:
What about `-> (Context<'t>, Result<Expression, Vec<Error>>)` instead? Same for `outer_statement`.

Reply:
```rust
285 |     let (ctx, value) = expression(ctx)?;
    |                        ^^^^^^^^^^^^^^^^ the `?` operator cannot be applied to type `(syntree::Context<'_>, std::result::Result<Expression, Vec<error::Error>>)`
``` 
I think it's worth keeping it like it is for egonomics with the `? ` operator. But it was a good idea!